File: python/clrun/scp/session.py
"""Handle input for SCP-backed sessions: transition + get_cli, append to buffer."""

from __future__ import annotations

from clrun.scp.client import (
    SCPClient,
    format_cli_for_buffer,
    resolve_input_to_action_or_body,
)
from clrun.buffer.buffer_manager import append_to_buffer
from clrun.pty.pty_manager import update_session


def handle_scp_input(
    terminal_id: str,
    text: str,
    project_root: str,
    base_url: str,
    run_id: str,
) -> dict:
    """Process input for an SCP session; returns dict with 'output' and optional 'done'."""
    client = SCPClient(base_url)
    cli = client.get_cli(run_id)
    options = cli.get("options") or []
    resolved = resolve_input_to_action_or_body(text, options)

    if not resolved:
        lines = ["Unknown option. Choose one of:"]
        for i, o in enumerate(options):
            label = o.get("label", o.get("action", ""))
            action = o.get("action", "")
            lines.append(f"  {i + 1}. {label} (or \"{action}\")")
        out = "\n".join(lines) + "\n"
        append_to_buffer(terminal_id, out, project_root)
        return {"output": out}

    action = resolved["action"]
    body = resolved.get("body")

    try:
        client.transition(action, body, run_id)
    except Exception as e:
        err_msg = str(e)
        out = f"Error: {err_msg}\n"
        append_to_buffer(terminal_id, out, project_root)
        return {"output": out}

    new_cli = client.get_cli(run_id)
    out = format_cli_for_buffer(new_cli)
    append_to_buffer(terminal_id, out, project_root)

    frame = client.get_frame(run_id)
    status = frame.get("status", "")
    if status in ("completed", "failed"):
        update_session(
            terminal_id,
            {"status": "exited", "last_exit_code": 0 if status == "completed" else 1},
            project_root,
        )
        return {"output": out, "done": True}
    return {"output": out}

File: src/scp/session.ts
import { SCPClient, formatCliForBuffer, resolveInputToActionOrBody } from './client';
import { appendToBuffer } from '../buffer/buffer-manager';
import { updateSession } from '../pty/pty-manager';

/**
 * Handle input for an SCP-backed session: resolve to transition, call API, append new CLI state to buffer.
 */
export async function handleScpInput(
  terminalId: string,
  input: string,
  projectRoot: string,
  baseUrl: string,
  runId: string
): Promise<{ output: string; done?: boolean }> {
  const client = new SCPClient(baseUrl);
  const cli = await client.getCli(runId);
  const options = cli.options ?? [];
  const resolved = resolveInputToActionOrBody(input, options);

  if (!resolved) {
    const lines: string[] = ['Unknown option. Choose one of:'];
    options.forEach((o, i) => lines.push(`  ${i + 1}. ${o.label} (or "${o.action}")`));
    const out = lines.join('\n') + '\n';
    appendToBuffer(terminalId, out, projectRoot);
    return { output: out };
  }

  try {
    await client.transition(resolved.action, resolved.body, runId);
  } catch (e) {
    const errMsg = e instanceof Error ? e.message : String(e);
    const out = `Error: ${errMsg}\n`;
    appendToBuffer(terminalId, out, projectRoot);
    return { output: out };
  }

  const newCli = await client.getCli(runId);
  const out = formatCliForBuffer(newCli);
  appendToBuffer(terminalId, out, projectRoot);

  const frame = await client.getFrame(runId);
  const isTerminal = frame.status === 'completed' || frame.status === 'failed';
  if (isTerminal) {
    updateSession(terminalId, { status: 'exited', last_exit_code: frame.status === 'completed' ? 0 : 1 }, projectRoot);
    return { output: out, done: true };
  }
  return { output: out };
}
